wdl-lint/src/rules/runtime_section_keys.rs

```diff
@@ -0,0 +1,501 @@
+            }
+            VisitReason::Exit => {
+                // If a runtime section span has not been encountered, then
+                // presumably there won't be any place to put missing
+                // recommended runtime keys.
+                let runtime_span = match self.current_runtime_span {
+                    Some(span) => span,
+                    None => return,
+                };
+
+                // SAFETY: at this point in the execution rule, the document
+                // version should have been encountered and set (for the outer
+                // [`Option`]).
+                if let Some(SupportedVersion::V1(minor_version)) =
+                    self.version.expect("the document version should be known")
+                {
+                    let specification = format!("the {minor_version} specification");
+
+                    if !self.non_reserved_keys.is_empty() {
+                        state.add(report_non_reserved_runtime_keys(
+                            &self.non_reserved_keys,
+                            runtime_span,
+                            &specification,
+                        ));
+                    }
+
+                    let recommended_keys = match minor_version {
+                        V1::Zero => recommended_keys(keys_v1_0()),
+                        V1::One => recommended_keys(keys_v1_1()),
+                        _ => return,
+                    };
+
+                    let missing_keys = recommended_keys
+                        .filter(|(key, _)| {
+                            !self.encountered_keys.iter().any(|s| s.as_str() == *key)
+                        })
+                        .map(|(key, _)| key)
+                        .collect::<Vec<_>>();
+
+                    if !missing_keys.is_empty() {
+                        state.add(report_missing_recommended_keys(
+                            missing_keys,
+                            runtime_span,
+                            &specification,
+                        ));
+                    }
+                }
+            }
+        }
+    }
+
+    fn runtime_section(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        section: &RuntimeSection,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        match self.current_runtime_span {
+            // Since validation and linting occur simultaneously, it is possible
+            // that we would hit this condition for a task with two `runtime`
+            // sections. Here, we simply ignore any `runtime` section but the
+            // first one, as multiple `runtime` sections will be flagged as an
+            // error by the validation.
+            Some(_) => {}
+            None => self.current_runtime_span = Some(span_of(section)),
+        }
+    }
+
+    fn runtime_item(&mut self, state: &mut Self::State, reason: VisitReason, item: &RuntimeItem) {
```

Comment:
What happens if someone does this?
```
task foo {
  .... 
  runtime {
    cromwell: {
      # cromwell specific 
      ...
    }
    miniwdl: {
      # miniwdl specific
      ...
    }
  }
}
```
That's an example in the spec (https://github.com/openwdl/wdl/blob/1.1.1/SPEC.md#conventions-and-best-practices).

Reply:
That's a good question. The specification is clear that any non-reserved keys are deprecated:

> 🗑 Additional arbitrary attributes may be specified in the runtime section, but these may be ignored by the execution engine. These non-standard attributes are called "hints". The use of hint attributes in the runtime section is deprecated; a later version of WDL will introduce a new hints section for arbitrary attributes and disallow non-standard attributes in the runtime section.

That said, the example does still exist in the specification. My feeling is that you should be able to `#@ except` those entries.

Reply:
> That said, the example does still exist in the specification. My feeling is that you should be able to `#@ except` those entries.

Sounds fine to me. I ofc have not memorized the docs for all the execution engines out there, but I don't recall ever actually seeing anyone adopt this convention. I think it's a weird hypothetical that most likely doesn't belong in the spec in the first place.